lib/prelude.dx

```diff
@@ -805,12 +851,73 @@ Dex's PRNG system is modelled directly after [JAX's](https://github.com/google/j
 '### Key functions
 
 -- TODO: newtype
-Key = Int64
-
-def hash (x:Key) (y:Int32) : Key = unsafeIO do
-  y64 = IToI64 y
-  %ffi threefry2x32 Int64 x y64
-def newKey (x:Int) : Key = hash (IToI64 0) x
+Key = Word64
+
+def threeFry2x32C (k:Int64) (count:Int64) : Int64 = unsafeIO do
+  %ffi threefry2x32 Int64 k count
+
+def applyRound (x:Word32) (y:Word32) (rot:Int) : (Word32&Word32) =
+  x = x + y
+  y = (y << rot) .|. (y >> (32 - rot))
+  y = x ^ y
+  (x, y)
+
+def threeFry2x32Dex (k:Word64) (count:Word64) : Word64 = 
+  -- Based on jax's threefry_2x32 by Matt Johnson and Peter Hawkins
+  rotations1 = [13, 15, 26, 6]
+  rotations2 = [17, 29, 16, 24]
+
+  k0 = W64ToW32 (k >> 32)
+  k1 = W64ToW32 k
+  -- TODO: add a fromHex?
+  k2 = k0 ^ k1 ^ (IToW32 466688986) -- 0x1BD11BDA
+
+  x = W64ToW32 (count >> 32)
+  y = W64ToW32 count
+  x = x + k0
+  y = y + k1
+
+  (x, y) = yieldState (x, y) \ref. for i.
```

Comment:
I could roll up these 5 loops into one loop, not sure what effect that would have (if any) on compile/run time?

Reply:
It would definitely lower compile time. I expect the run-time would be pretty much the same because this looks like something LLVM should be perfectly capable of optimizing. Rolling into a loop sounds great to me.